Suppose that a test for using a particular drug is 97% sensitive and 95% specific.

That is, the test will produce 97% true positive results for drug users and 95% true negative results for non-drug users.

Suppose that 0.5% of the general population are users of the drug. What is the probability that a randomly selected individual with a positive test is a drug user?

We will write a custom function which accepts the test capabilities and the prior knowledge of drug user percentage as input and produces the output probability of a test-taker being an user based on a positive result.


The Bayes' rule based calculation is given by the following equation,

P(User|+)
         = [P(+|User).P(User)]/P(+)

         = [P(+|User).P(User)] / [P(+|User).P(User)+P(+|Non-user).P(Non-user)]

Here,
P(User)=Prevelance rate
P(Non-user)=1 − Prevelance rate
P(+|User)= Sensitivity
P(−|Non-user) = Specificity
P(+|Non-user)=1 − Specificity

In [1]:
def drug_user(prob_th=0.5,sensitivity=0.99,specificity=0.99,prevelance=0.01,verbose=True):
    """
    """
    p_user = prevelance
    p_non_user = 1-prevelance
    p_pos_user = sensitivity
    p_neg_user = specificity
    p_pos_non_user = 1-specificity

    num = p_pos_user*p_user
    den = p_pos_user*p_user+p_pos_non_user*p_non_user

    prob = num/den

    if verbose:
        if prob > prob_th:
            print("The test-taker could be an user")
        else:
            print("The test-taker may not be an user")

    return prob


# Run with the details given in the original question

p = drug_user(prob_th=0.5,sensitivity=0.97,specificity=0.95,prevelance=0.005)
print("Probability of the test-taker being a drug user is:",round(p,3))

The test-taker may not be an user
Probability of the test-taker being a drug user is: 0.089


How about applying the Bayes' rule twice (or more)?

The best thing about Bayesian inference is the ability to use prior knowledge in the form of Prior probability term in the numerator of the Bayes' theorem.

In this setting of drug screening, the prior knowledge is nothing but the computed probability of a test which is then fed back to a next test.

That means, for these cases, where the prevelance rate in the general population is extremely low, one way to increase confidence is to prescribe subsequent test if the first test result is positive.

The probability from the first test becomes the Prior for the second test i.e. the P(User)

is not the general prevelance rate anymore for this second test, but the probability from the first test.

Let's see how the probability evolves with a series of tests...

In [2]:
p1 = drug_user(prob_th=0.5,sensitivity=0.97,specificity=0.95,prevelance=0.005)
print("Probability of the test-taker being a drug user, in the first round of test, is:",round(p1,3))
print()
p2 = drug_user(prob_th=0.5,sensitivity=0.97,specificity=0.95,prevelance=p1)
print("Probability of the test-taker being a drug user, in the second round of test, is:",round(p2,3))
print()
p3 = drug_user(prob_th=0.5,sensitivity=0.97,specificity=0.95,prevelance=p2)
print("Probability of the test-taker being a drug user, in the third round of test, is:",round(p3,3))

The test-taker may not be an user
Probability of the test-taker being a drug user, in the first round of test, is: 0.089

The test-taker could be an user
Probability of the test-taker being a drug user, in the second round of test, is: 0.654

The test-taker could be an user
Probability of the test-taker being a drug user, in the third round of test, is: 0.973


**Example 1:** A person has undertaken a job. The probabilities of completion of the job on time with and without rain are 0.44 and 0.95 respectively. If the probability that it will rain is 0.45, then determine the probability that the job will be completed on time.

Formulate this problem and code in python.
(Reference: https://www.geeksforgeeks.org/bayes-theorem/)

Solution :

Probability of job completion on time with rain: P(C|R) = 0.44
Probability of job completion on time without rain: P(C|R') = 0.95
Probability of rain: P(R) = 0.45

To Find:
        Probability of job completion on time: P(C)

In [3]:
def job_completion_prob(p_c_r, p_c_not_r, p_r):
    """
    Calculates the probability of job completion on time.

    Args:
        p_c_r: Probability of job completion given rain.
        p_c_not_r: Probability of job completion given no rain.
        p_r: Probability of rain.

    Returns:
        The probability of job completion on time.
    """

    p_c = p_c_r * p_r + p_c_not_r * (1 - p_r)
    return p_c


In [4]:

p_c = job_completion_prob(0.44, 0.95, 0.45)
print("Probability of job completion on time:", p_c)

Probability of job completion on time: 0.7204999999999999


**Example 2:** Now update the probality of job completion with rain in **example 1** to the one obtained by solving example 1. Follow the same procedure above and calculate the probability that the job will be completed on time if the probability of rain is 0.55.

In [7]:
# Calculate the initial probability of job completion
p_c_initial = job_completion_prob(0.44, 0.95, 0.45)


In [8]:
# Update the probability of job completion given rain
p_c_r_updated = p_c_initial


In [9]:
# Calculate the new probability of job completion with p_r = 0.55
p_c_new = job_completion_prob(p_c_r_updated, 0.95, 0.55)

print("Updated probability of job completion:", p_c_new)

Updated probability of job completion: 0.8237749999999999


This means that with the updated probability of job completion given rain (based on the initial calculation), and a rain probability of 0.55, the overall probability of job completion on time is approximately 0.824.